libnmstate/nm/device.py

```diff
@@ -49,6 +49,68 @@ def delete(dev):
         con_profile.delete()
 
 
+def reapply(dev, connection_profile):
+    mainloop = nmclient.mainloop()
+    mainloop.push_action(_safe_reapply_async, dev, connection_profile)
+
+
+def _safe_reapply_async(dev, connection_profile):
+    mainloop = nmclient.mainloop()
+    cancellable = mainloop.new_cancellable()
+
+    if _wait_for_active_connection_async(dev, connection_profile):
+        return
+
+    version_id = 0
+    flags = 0
+    user_data = mainloop, dev, cancellable
+    dev.reapply_async(
+        connection_profile,
+        version_id,
+        flags,
+        cancellable,
+        _reapply_callback,
+        user_data,
+    )
+
+
+def _wait_for_active_connection_async(dev, connection_profile):
+    active_conn = connection.get_device_active_connection(dev)
+    if active_conn:
+        act_conn = ac.ActiveConnection(active_conn)
+        if act_conn.is_activating:
+            logging.debug(
+                'Connection activation in progress: dev=%s, state=%s',
+                act_conn.devname, act_conn.state)
+            conn = connection.ConnectionProfile(connection_profile)
+            conn.waitfor_active_connection_async(act_conn)
+            return True
+    return False
+
+
+def _reapply_callback(src_object, result, user_data):
+    mainloop, nmdev, cancellable = user_data
+    mainloop.drop_cancellable(cancellable)
+
+    devname = src_object.get_iface()
+    try:
+        success = src_object.reapply_finish(result)
+    except Exception as e:
+        if mainloop.is_action_canceled(e):
+            logging.debug('Device reapply aborted on %s: error=%s', devname, e)
+        else:
+            mainloop.quit(
+                'Device reapply failed on {}: error={}'.format(devname, e))
```

Comment:
Inside the mainloop we cannot raise.
The mainloop needs to be gracefully terminated, otherwise only the timeout will end it (and in a non graceful manner).

Perhaps investigating exceptions inside the mainloop is to be investigated further, we should consider it.

Reply:
https://nmstate.atlassian.net/browse/NMSTATE-208